### ISSUE TITLE
Consulting about HD loss

### ISSUE THREAD

**Issue by abcxubu** [SilmarilBearer/HausdorffLoss#4]
Sorry for bothering you again. Recently, based on TransUnet, I am doing vessel image segmentation (the extreme imbalanced problem, the example of the ground truth is attached). I tried used the Hoausdorff Loss that you provided (total loss = loss_HD + alpha * loss_Dice, where alpha = mean(loss_HD)/mean(loss_Dice)) as described in the original paper. 
The training loss curve is attached. In the first iteration, loss_HD = 4534 and loss_Dice = 0.55. In the second iteration, loss_HD = 0.0058 and loss_Dice = 0.5. In the third iteration, loss_HD = 0.0018 and loss_Dice = 0.5. In the training, I found that from the second iteration, all the prediction results are the background (i.e. the images are black, and it did not find the vessel). And in the testing process, all the prediction results are the background.
I have tried my best but still can not find where I am wrong. Could you please give me some advice? Is it because total loss can't handle extreme imbalance problem? Thanks.
![CHAI CONG JUAN_00219_098 10-gt](https://user-images.githubusercontent.com/25636252/124931325-aff62f00-e034-11eb-87f7-eaeb84ddfc46.png)
![CHAI CONG JUAN_00079_035 10-gt](https://user-images.githubusercontent.com/25636252/124931331-b1bff280-e034-11eb-9773-7c2e58fdab7d.png)
![Loss](https://user-images.githubusercontent.com/25636252/124931339-b2f11f80-e034-11eb-9aed-161f0b326e80.PNG)

**Comment by PatRyg99** [comment]
I believe that's a common problem for distance based losses. They tend to behave unstable when net returns random results (as it is for the first few steps) and label everything as background when classes are highly imbalanced.

I think the problem may lay in your alpha change being dependent on loss value thus being very unstable in early stages. I would try to define loss as: loss = dice * alpha + Haussdorf * (1 - alpha), and start for example from alpha=0.95 and define it to be 0.5 or even lower at the training end - alpha will progressively go down during training time. This way you allow dice do the heavy lifting early to establish some reasonable prediction and not allow early Haussdorf anomalies confuse the network that much. Later on when net outputs better predictions Haussdorf loss should be stable.

**Comment by abcxubu** [comment]
Thanks a lot for your advice. I have used the trick that you provided. The initial alpha is set to 0.00001, and the alpha increases by 0.00002 for every iteration. Then the network can find the vessels in the training and testing. The loss curve is attached. The loss is a little unstable, and I think I need to give a better initial alpha and a better strategy to change the alpha.
![new_loss](https://user-images.githubusercontent.com/25636252/125193410-e58e5880-e27e-11eb-87fc-0d881610188c.PNG)
